### PR TITLE
GCS_MAVLink: home position should always send

### PR DESCRIPTION
**Pull request by magicrub**
When a GCS requests the home position and we respond via ```send_home_position()``` we used to always send it if home was set. Now we'll ignore the request for many more seconds until ```home.get_vector_from_origin_NEU()``` returns true which is when the EKF origin is set. Is that really necessary? It's just populating another part of the home_position() message which is otherwise 0 and used to be hard-coded to 0. If a GCS really wants that info then it can re-request until it gets non-zero data there.

### CODE REVIEW

**Comment by rmackay9** [comment]
I've always thought that there cannot be a home position until the EKF origin has been set so I'm not sure the current code needs to be changed.

**Comment by magicrub** [comment]
well, it used to be once you get a GPS lock a GCS could ask you what your home location is. Now it takes a dozen seconds later and meanwhile swallows up your request without a NACK or anything

**Comment by magicrub** [comment]
your home isn't what's changing.. it's the origin that does and it's just adding to the message. The home is valid while it's rejecting the request... and frankly a GCS doesn't care about the EKF origin in most cases

**Comment by magicrub** [comment]
ping

**Comment by magicrub** [comment]
This likely deserves a discussion at a dev call

**Comment by magicrub** [comment]
PR that caused this new behavior: https://github.com/ArduPilot/ardupilot/pull/21642
MAVLink msg: https://mavlink.io/en/messages/common.html#HOME_POSITION

Outcome from devcall discussion: The field should really be NaN as default instead of 0. Same goes for q[4]

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 1]
```diff
@@ -2749,7 +2749,7 @@ void GCS_MAVLINK::send_home_position() const
     // get home position from origin
     Vector3f home_pos_neu_cm;
     if (!home.get_vector_from_origin_NEU(home_pos_neu_cm)) {
-        return;
+        home_pos_neu_cm.zero();
```
Comment (tridge) [inline libraries/GCS_MAVLink/GCS_Common.cpp:2752]: should be nan, and probably for q and offsets too

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by magicrub** [comment]
@peterbarker said he'll contact @hamishwillee regarding MAVLink changes

**Comment by hamishwillee** [comment]
He did. Makes a lot of sense. 

Regarding invalid values, by convention we use zero "if 0 is invalid in the context of the field" - I am assuming all zeros is an invalid quaternion. 
Otherwise we use a max field value for ints/uints and NaN for floats.

Consider also making the Lat/Lon/Alt as NaN invalid too, if it is possible the message will be requested when you have absolute but no local position. Also it is possible this might be streamed at low rate in some flight stacks, when there is no home position, or only some data is available.

Lastly, if you get a request and you have no home position, NACK the request with MAV_RESULT_TEMPORARILY_REJECTED and don't send the message. 

Can you PR the changes upstream please.

**Comment by hamishwillee** [comment]
Actually there is already a PR to specify quaternions as NaN - @auturgy approved it, and I have merged it https://github.com/mavlink/mavlink/pull/1843

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 1 above)
Comment (magicrub) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Q and InvalidOrigin is now NAN

**Comment by magicrub** [comment]
I've fixed this per @tridge 's comments and did it in two commits so we can drop the Home one if necessary.
- Q[4] = NaNs via upstream mavlink spec change https://github.com/mavlink/mavlink/pull/1843
- Invalid Home as NaN which is not specified in mavlink and thus potential diverges from the spec with potential negative effects in GCSs

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 2]
```diff
@@ -2761,19 +2761,23 @@ void GCS_MAVLINK::send_home_position() const
 
     // get home position from origin
     Vector3f home_pos_neu_cm;
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Common.cpp:2763]:
```suggestion
    Vector3f home_pos_ned;
```

You're nicely converting it immediately after it is fetched, so the name is almost immediately made incorrect :-)

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 2 above)
Comment (magicrub) [reply libraries/GCS_MAVLink/GCS_Common.cpp:2763]: Ha! I didn't even notice. I'll change that.

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 3]
```diff
@@ -2761,19 +2761,23 @@ void GCS_MAVLINK::send_home_position() const
...
+    if (home.get_vector_from_origin_NEU(home_pos_neu_cm)) {
+        // convert NEU in cm to NED in meters
+        home_pos_neu_cm *= 0.01f;
+        home_pos_neu_cm.z *= -1;
+    } else {
+        home_pos_neu_cm = Vector3f(NAN, NAN, NAN);
     }
 
-    const float q[4] = {1.0f, 0.0f, 0.0f, 0.0f};
+    const float q[4] = {NAN, NAN, NAN, NAN};
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Common.cpp:2772]:
```suggestion
    const float q[4] {NAN, NAN, NAN, NAN};
```

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 4]
```diff
@@ -2761,19 +2761,23 @@ void GCS_MAVLINK::send_home_position() const
...
     // get home position from origin
     Vector3f home_pos_neu_cm;
-    if (!home.get_vector_from_origin_NEU(home_pos_neu_cm)) {
-        return;
+    if (home.get_vector_from_origin_NEU(home_pos_neu_cm)) {
+        // convert NEU in cm to NED in meters
+        home_pos_neu_cm *= 0.01f;
+        home_pos_neu_cm.z *= -1;
+    } else {
+        home_pos_neu_cm = Vector3f(NAN, NAN, NAN);
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Common.cpp:2769]:
```suggestion
        home_pos_neu_cm = Vector3f{NAN, NAN, NAN};
```

**Review by peterbarker — APPROVED** [review]

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 2 above)
Comment (magicrub) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: done

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 1 above)
Comment (magicrub) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: fixed

**Review by peterbarker — APPROVED** [review]

**Pull request merged**